[FLINK-36809][table] Support ignoreIfExists param for createTable

gustavodemorais opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Support ignoreIfExists for createTable. Works exactly the same way for other resources, like createFunction. Similar to SQL's `IF NOT EXISTS`.


## Brief change log

  - Added one more signature for TableEnvironment's createTable supporting ignoreIfExists
  - Added testCreateTableIfNotExistsFromDescriptor


## Verifying this change
This change added tests and can be verified as follows:

  - Added test that validates that running createTable, with ignoreIfExists set to true, for an existing table leads to a no op. Also checks it doing the same ignoreIfExists set to false throws an exception.

## Does this pull request potentially affect one of the following parts:

  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - Dependencies (does it add or upgrade a dependency): no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -745,6 +745,35 @@ void createTemporarySystemFunction(
+     * @param ignoreIfExists If a table exists under the given path and this flag is set, no
+     *     operation is executed. An exception is thrown otherwise.
+     */
+    void createTable(String path, TableDescriptor descriptor, boolean ignoreIfExists);
```
can you check whether Python already supports `createTable` with TableDescriptors. if yes, we should also add the method there. check `flink/flink-python/pyflink/table/table_environment.py`

gustavodemorais replied:
Yes, it does. I'm getting my python environment set up right

twalthr commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/api/TableEnvironmentTest.java:
```diff
@@ -100,6 +101,55 @@ void testCreateTableFromDescriptor() throws Exception {
+                                                .option("a", "Test")
+                                                .build(),
+                                        false))
+                .isInstanceOf(ValidationException.class);
```
ideally, we should also check the error message. at least parts of it. `hasMessageContaining` could do the job?

twalthr commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/api/TableEnvironmentTest.java:
```diff
@@ -100,6 +101,55 @@ void testCreateTableFromDescriptor() throws Exception {
+        final String catalog = tEnv.getCurrentCatalog();
+        final String database = tEnv.getCurrentDatabase();
+
+        final Schema schema = Schema.newBuilder().column("f0", DataTypes.INT()).build();
+        tEnv.createTable(
+                "T",
+                TableDescriptor.forConnector("fake").schema(schema).option("a", "Test").build(),
+                true);
+
+        final ObjectPath objectPath = new ObjectPath(database, "T");
+        assertThat(
+                        tEnv.getCatalog(catalog)
+                                .orElseThrow(AssertionError::new)
+                                .tableExists(objectPath))
+                .isTrue();
+
+        final CatalogBaseTable catalogTable =
+                tEnv.getCatalog(catalog).orElseThrow(AssertionError::new).getTable(objectPath);
+        assertThat(catalogTable).isInstanceOf(CatalogTable.class);
+        assertThat(catalogTable.getUnresolvedSchema()).isEqualTo(schema);
+        assertThat(catalogTable.getOptions())
+                .contains(entry("connector", "fake"), entry("a", "Test"));
```
this code is duplicated with the test above. create a private helper method e.g. `assertTableFromDescriptor(String name)`

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -718,6 +718,36 @@ void createTemporarySystemFunction(
+     *     .build())
+     *   .option(DataGenOptions.ROWS_PER_SECOND, 10)
+     *   .option("fields.f0.kind", "random")
+     *   .build());
```
update example

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -745,6 +775,35 @@ void createTemporarySystemFunction(
+     *     .build())
+     *   .option(DataGenOptions.ROWS_PER_SECOND, 10)
+     *   .option("fields.f0.kind", "random")
+     *   .build());
```
update example

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -745,6 +745,35 @@ void createTemporarySystemFunction(
+     * in the catalog.
+     *
+     * <p>If the table should not be permanently stored in a catalog, use {@link
+     * #createTemporaryTable(String, TableDescriptor)} instead.
```
update link

gustavodemorais replied:
> update link
ops, thanks!

twalthr reviewed (commented):
The Java side should be good in the next iteration.

davidradl commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -718,6 +718,36 @@ void createTemporarySystemFunction(
+     * <p>The {@link TableDescriptor descriptor} is converted into a {@link CatalogTable} and stored
+     * in the catalog.
+     *
+     * <p>Temporary objects can shadow permanent ones. If a permanent object in a given path exists,
```
I read sentence "Temporary objects can shadow permanent ones"
- I am not sure why I would want to do this ; is this shadowing capability documented?
- I suggest this sentence use `table `instead of` object`?

gustavodemorais replied:
- https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/common/#temporary-vs-permanent-tables
- I agree that `tables` instead of `objects` would be easier to digest in this context. On the other hand, the temporary objects concept seems to exist for multiple resources, like views, and having the same sentence might help people recognize this as being the same pattern. I have no strong opinion here though.

davidradl commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -718,6 +718,36 @@ void createTemporarySystemFunction(
+     * in the catalog.
+     *
+     * <p>Temporary objects can shadow permanent ones. If a permanent object in a given path exists,
+     * it will be inaccessible in the current session. To make the permanent object available again
```
for sentences `If a permanent object in a given path exists, it will be inaccessible in the current session. To make the permanent object available again one can drop the corresponding temporary object.` 
- can we use `table` or `TableDescriptor ` instead of `object`?
- is this true if ignoreIfExists is set.

gustavodemorais replied:
- Same as https://github.com/apache/flink/pull/25704#discussion_r1862771988
- Good question. I'm also not sure and would need to double-check

davidradl commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -718,6 +718,36 @@ void createTemporarySystemFunction(
+     * @param path The path under which the table will be registered. See also the {@link
+     *     TableEnvironment} class description for the format of the path.
+     * @param descriptor Template for creating a {@link CatalogTable} instance.
+     * @param ignoreIfExists If a table exists under the given path and this flag is set, no *
```
I am curious, if there is an existing table , temporary table , view and materialized table, does this flag act the same in each case.

gustavodemorais replied:
I think it does behave the same

davidradl commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/api/TableEnvironmentTest.java:
```diff
@@ -49,55 +50,78 @@ class TableEnvironmentTest {
+                                        TableDescriptor.forConnector("fake")
+                                                .schema(schema)
+                                                .option("a", "Test")
+                                                .build(),
```
I assume table T existing in this case - so no exception is thrown. I suggest adding a test to ensure that creating a temp table with the flag set means that the table will be created when there is no existing table.

gustavodemorais replied:
I've updated the tests so both `assertCreateTableFromDescriptor` and `assertTemporaryCreateTableFromDescriptor` helper functions have a n gnoreIfExists param and we test the case you asked for 👍

gustavodemorais commented:
@twalthr @davidradl thanks for the reviews 👍 I think I've addressed all comments, take a look. 

Updating the python api was a bit tricky with the setup (curious to see if lint will fail), but the implementation should be working. Tests are identical to the java implementation and are passing.

twalthr reviewed (commented):
Thanks for the update! The PR is in a very good shape now. I had some last comments to remove code deduplication and improve readability. Should be good to merge in the next iteration.

twalthr reviewed (approved):
LGTM. Given that CI shows green.
